votingsystem/src/voting.rs

```diff
@@ -0,0 +1,419 @@
+pub trait VoteManager: Service {
+    fn get_vote(&self, vote_id: &VoteId) -> Result<Vote, Error>;
+    fn get_shares(&self, vote_id: &VoteId) -> Result<u32, Error>;
+}
+
+impl VoteManager for Context {
+    fn get_vote(&self, vote_id: &VoteId) -> Result<Vote, Error> {
+        if !self.storage().has(vote_id.as_ref()) {
+            return Err(Error::InvalidVote)
+        }
+        let vote: Vote = serde_cbor::from_slice(&self.storage().get(vote_id.as_ref()).unwrap()).unwrap();
+        Ok((vote))
+    }
+
+    fn get_shares(&self, vote_id: &VoteId) -> Result<u32, Error> {
+        let vote: Vote = serde_cbor::from_slice(&self.storage().get(vote_id.as_ref()).unwrap()).unwrap();
+        let vote_paper_id = vote.corresponding_paper_id;
+        let vote_paper: VotePaper =
+            serde_cbor::from_slice(&self.storage().get(vote_paper_id.0.as_ref()).unwrap()).unwrap();
+        Ok(vote_paper.number_of_shares)
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TxCreateVotePaper {
+    general_meeting_id: GeneralMeetingId,
+    agenda_number: u32,
+    number_of_shares: u32,
+    voter_name: String,
+    voter_public_key: Public,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TxVote {
+    vote_paper_id: H256,
+    choice: String,
+    voter_signature: Signature,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum TxAction {
+    CreateVotePaper(TxCreateVotePaper),
+    Vote(TxVote),
+}
+
+impl Action for TxAction {}
+pub type OwnTransaction = crate::common::SignedTransaction<TxAction>;
+
+impl Context {
+    fn excute_tx(&mut self, transaction: &Transaction) -> Result<H256, ExecuteError> {
+        if transaction.tx_type() != "Vote" {
+            return Err(ExecuteError::InvalidMetadata)
+        }
+        let tx: OwnTransaction =
+            serde_cbor::from_slice(&transaction.body()).map_err(|_| ExecuteError::InvalidFormat)?;
+        tx.verify().map_err(|_| ExecuteError::InvalidSign)?;
+
+        match tx.tx.action {
+            TxAction::CreateVotePaper(TxCreateVotePaper {
+                general_meeting_id,
+                agenda_number,
+                number_of_shares,
+                voter_name: voter_name,
+                voter_public_key,
+            }) => {
+                let meeting: GeneralMeeting = self.meeting_mut().get_meeting(&general_meeting_id).unwrap();
+                if meeting.id != general_meeting_id {
+                    return Err(ExecuteError::InvalidMeeting)
+                }
+                if meeting.tallying_time.time < self.block_header.as_ref().unwrap().timestamp() {
+                    return Err(ExecuteError::InvalidMeetingTime)
+                }
+                if meeting.number_of_agendas < agenda_number {
+                    return Err(ExecuteError::InvalidNumberofAgendas)
+                }
+                if meeting.voting_end_time.time > self.block_header.as_ref().unwrap().timestamp() {
+                    return Err(ExecuteError::InvalidVotingTime)
```

Comment:
Please the name of the error.
`VotingTime` is correct. The problem is that `CreateVotePaper` transaction is included in a block too late.